### ISSUE TITLE
Value injection, structs and constructors don't work together

### ISSUE THREAD

**Issue by ghost91-** [mbierlee/poodinis#25]
Consider the following example:
```D
import poodinis;

struct TestStruct
{
}

class TestStructInjector : ValueInjector!TestStruct
{
	TestStruct get(string key)
	{
		return TestStruct.init;
	}
}

class TestClass
{
	this(TestStruct testStruct)
	{
		this.testStruct = testStruct;
	}

	@Value("") TestStruct testStruct;
}

void main()
{
	auto container = new shared DependencyContainer();
	container.register!(ValueInjector!TestStruct, TestStructInjector);
	container.register!(TestClass);
}
```
This does not compile:
```
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/container.d(169,39): Error: cannot implicitly convert expression typeid(TestStruct) of type object.TypeInfo_Struct to object.TypeInfo_Class
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/container.d(181,36): Error: template instance `AutowiredRegistration!(TestStruct)` does not match template declaration AutowiredRegistration(RegistrationType : Object)
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/container.d(305,59): Error: template instance `poodinis.container.DependencyContainer.register!(TestStruct, TestStruct)` error instantiating
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/container.d(266,60):        instantiated from here: resolve!(TestStruct, TestStruct)
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/factory.d-mixin-152(153,63):        instantiated from here: resolve!(TestStruct)
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/container.d(180,36):        instantiated from here: ConstructorInjectingInstanceFactory!(TestClass)
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/container.d(147,53):        instantiated from here: register!(TestClass, TestClass)
source/app.d(29,11):        instantiated from here: register!(TestClass)
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/container.d(333,68): Error: cannot cast expression registration.getInstance(new AutowireInstantiationContext) of type object.Object to TestStruct
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/container.d(338,68): Error: cannot cast expression registration.getInstance(autowireContext) of type object.Object to TestStruct
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/container.d(321,41): Error: template instance `poodinis.container.DependencyContainer.resolveAutowiredInstance!(TestStruct)` error instantiating
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/container.d(266,60):        instantiated from here: resolve!(TestStruct, TestStruct)
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/factory.d-mixin-152(153,63):        instantiated from here: resolve!(TestStruct)
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/container.d(180,36):        instantiated from here: ConstructorInjectingInstanceFactory!(TestClass)
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/container.d(147,53):        instantiated from here: register!(TestClass, TestClass)
source/app.d(29,11):        instantiated from here: register!(TestClass)
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/container.d(312,24): Error: cannot implicitly convert expression null of type typeof(null) to TestStruct
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/factory.d(163,44): Error: no property create for type object.TypeInfo_Struct
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/container.d(180,36): Error: template instance `poodinis.factory.ConstructorInjectingInstanceFactory!(TestStruct)` error instantiating
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/container.d(305,59):        instantiated from here: register!(TestStruct, TestStruct)
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/container.d(266,60):        instantiated from here: resolve!(TestStruct, TestStruct)
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/factory.d-mixin-152(153,63):        instantiated from here: resolve!(TestStruct)
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/container.d(180,36):        ... (1 instantiations, -v to show) ...
../../.dub/packages/poodinis-8.0.2/poodinis/source/poodinis/container.d(147,53):        instantiated from here: register!(TestClass, TestClass)
source/app.d(29,11):        instantiated from here: register!(TestClass)
/usr/bin/dmd failed with exit code 1.
```
It works correctly when using basic types (e.g. `int`) instead of `TestStruct` and it also works correctly when removing the constructor in `TestClass`.

**Comment by ghost91-** [comment]
The concrete usecase is that I'd like to avoid using poodinis to wire my components in unittests. It would be a lot less boilerplate, to just pass the dependencies as parameters to a constructor, because a lot of them will be mocked anyways, so wiring is not neccessary. So I'd be ok with that constructor simply being skipped.

**Comment by mbierlee** [comment]
Fixed in v8.0.3
